### PR TITLE
Split librustc::{traits,infer} to a separate crate rustc_infer

### PR DESCRIPTION
**Pull request by cjgillot**
This is still very much work in progress.
Three functions are between dimensions (at the end of `rustc::traits`), waiting for some dependency breaking scheme.
Please tell me if the approach seems sound, and how you would like to split this PR up.

The formatting is deliberately off, to ease rebasing. 

cc #65031

### CODE REVIEW

**Comment by cjgillot** [comment]
Once #67780 and #67797 are merged, the `resolve_instance` query will be movable to a downstream crate. When that is done, this PR will be ready.

File: src/librustc_infer/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+[dependencies]
+arena = { path = "../libarena" }
+bitflags = "1.2.1"
+fmt_macros = { path = "../libfmt_macros" }
+graphviz = { path = "../libgraphviz" }
+jobserver = "0.1"
+num_cpus = "1.0"
+scoped-tls = "1.0"
+log = { version = "0.4", features = ["release_max_level_info", "std"] }
+rustc-rayon = "0.3.0"
```
Comment (Zoxc) [inline src/librustc_infer/Cargo.toml]: You could strip all these and only add back the used dependencies.

File: src/librustc_typeck/check/demand.rs [hunk 2]
```diff
@@ -797,9 +797,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         );
                         err.warn(
                             "if the rounded value cannot be represented by the target \
-                                  integer type, including `Inf` and `NaN`, casting will cause \
-                                  undefined behavior \
-                                  (https://github.com/rust-lang/rust/issues/10184)",
+                             integer type, including `Inf` and `NaN`, casting will cause \
```
Comment (Zoxc) [inline src/librustc_typeck/check/demand.rs]: What's up with all these formatting changes?

**Comment by cjgillot** [comment]
Pushed a cleaner and rebased version. It contains some other PRs as merge commits. Those ought to be rebased out.

File: src/librustc_typeck/check/demand.rs (hunk 2 above)
Comment (cjgillot) [reply src/librustc_typeck/check/demand.rs]: A badly configured editor. Reverted.

**Comment by Zoxc** [comment]
Nominating for @rust-lang/compiler. This PR mostly moves code around to enable splitting out `rustc::traits` and `rustc::infer` to a new crate `rustc_infer`. Some methods are converted to trait extensions methods, but I think quite a few of those could be made free standing functions. This would help quite a bit towards cutting down the size of `librustc` and it was done relatively cleanly. I was expecting it to be harder. Just checking in to see if anyone has some concerns about doing this.

**Comment by eddyb** [comment]
I don't like the name `rustc_infer` for holding `rustc::traits`, to the point where I'd prefer a second "`rustc_traits`" crate that depends on `rustc_infer`.

**Comment by Zoxc** [comment]
@eddyb I don't think splitting `rustc::traits` and `rustc::infer` into separate crates is going to be easy and I would consider extracting them from `rustc` a first step towards that. I do agree that `rustc_infer` is not a great name, since it can easily be confused with the existing `rustc::infer`.

**Comment by nikomatsakis** [comment]
Longer term, as part of chalkification, I would like to break both of these bits of code out and move them to a distinct library. But in the **shorter term**, it's always nice to have less code in rustc, so I think this makes sense in the abstract. I'll have to review the diffs I guess.

One thing I would like to push back on, though, is the implication that free-standing functions are preferable to extension traits. I've lately seen a trend towards converting things to free functions. Sometimes it's fine, or at least neutral, but I find that, for many methods, having a distinguished "subject" is quite helpful (e.g., I prefer `trait_ref.self_type()` to `self_type(trait_ref)`, or what have you). I guess that on this matter, like so many matters, I would prefer we don't try to take some kind of automatic hard line.

**Comment by Centril** [comment]
**The move to free functions in these cases is great in my view.** It makes the code overall clearer wrt. imports and whatnot. Also, I personally find extension traits to be confusing, using traits for something other than ad-hoc polymorphism, with a single implementation, and not for usage as a bound. These extension traits, because you only want one of them to reduce import boilerplate, also nudge you towards artificial groupings of mostly unrelated functions, reducing modularity and making it harder to move the unrelated functions to more appropriate places as time passes.

**However, I believe there's a missing language feature here.** We could support calling free functions which indicate that they allow being called with a receiver:

```rust
// module_a.rs:

fn my_method(self: MyType, other_param: u8) -> u8 { ... }

// b.rs:

use module_a::my_method;

fn main() {
    let recv = make();
    recv.my_method(42);
}
```

but supporting a new language feature is neither here nor there.

**Comment by nikomatsakis** [comment]
I don't disagree with your points, I guess I just don't experience them as that severe. I've never had the problem of artificially grouping methods together, for example, I feel like the extension traits tend to either be a logical bundle of methods or else be local to a file. (I do find it a bit repetitive to define the traits separately from the impl, though.) Obviously this is sort of a matter of a taste. 

(Regarding language extensions, that could be nice, not sure; one thing I've also considered to reduce boilerplate at least would be a proc-macro that lets you write an extension trait in one block, sort of like

```
#[extension_trait]
impl FooExt for Foo { ... }
```

which would create a `trait FooExt` and an impl.)

**Comment by cjgillot** [comment]
> I don't like the name `rustc_infer` for holding `rustc::traits`, to the point where I'd prefer a second "`rustc_traits`" crate that depends on `rustc_infer`.

What would you like to see in each crate? The infer and traits modules here are quite intertwined. I haven't seen an easy cut yet.

About naming, please tell me what you would rather see.

**Comment by eddyb** [comment]
> The infer and traits modules here are quite intertwined.

That doesn't *feel* right. `traits` definitely depends on `infer`, but how strong is the dependency the other way around? Could `infer` code that depends on `traits` be provided in `traits` instead?

**Comment by cjgillot** [comment]
> > The infer and traits modules here are quite intertwined.
> 
> That doesn't _feel_ right. `traits` definitely depends on `infer`, but how strong is the dependency the other way around? Could `infer` code that depends on `traits` be provided in `traits` instead?

I do not have a complete list. Until now I used the stupid method: comment out `mod traits` and fix compile errors.
However, there are quite a few of them. Some are just types, for instance the `*Obligation*` types. Some are routines, for instance `TyCtxt::resolve_regions_and_report_errors` -> `infer::OutlivesEnvironment::new` -> `traits::query::outlives_bounds::explicit_outlives_bounds`.

I am trying to see what can be split and how, but this may take some time.

**Comment by eddyb** [comment]
> Some are just types, for instance the `*Obligation*` types.

Presumably those would still need to be `rustc`.

<hr/>

> `infer::OutlivesEnvironment::new` -> `traits::query::outlives_bounds::explicit_outlives_bounds`

Hmm, `query` would make me think that goes through a query, which would break up the dependency.

<hr/>

I assume `TraitEngine` is related to the Chalk integration?
That's new to me, I'm less sure about that being decouplable.

**Comment by cjgillot** [comment]
Rebased on top of #69100, so we can avoid waiting for #67797.

**Comment by cjgillot** [comment]
Rebased. This PR is ready for review.

**Comment by Zoxc** [comment]
This looks good to me besides the dependency mentioned by @matthewjasper, but that can be fixed later in https://github.com/rust-lang/rust/pull/69076.

@bors r+ p=1

**Comment by Zoxc** [comment]
This seems to have regressed performance a bit: https://perf.rust-lang.org/compare.html?start=5e7af4669f80e5f682141f050193ab679afdb4b1&end=a643ee8d693b8100e6f54f2a01ff7cde05eb65c5&stat=instructions:u

My guess is that something hot doesn't get inlined in type checking anymore.

**Comment by eddyb** [comment]
Why does this keep duplicated definitions, e.g. `rustc::traits::project::ProjectionCache`?

**Comment by eddyb** [comment]
This PR made some fields `pub` that weren't before (e.g. in `SelectionCache`, `EvaluationCache` etc.).
That's an abstraction boundary break, IMO, but I'm not sure how I'd address it.

(if you saw a previous comment, please disregard, my editor was playing tricks on me)

**Comment by mati865** [comment]
@Zoxc maybe it's worth to open issue so this regression isn't forgotten. Seems like low hanging fruit.

**Pull request merged**